R/stat-sina.r

```diff
@@ -0,0 +1,186 @@
+#' @rdname geom_sina
+#' @inheritParams stat_identity
+#' @param binwidth Binning factor. The range of the values in \code{aes(y)}
+#' are binned in windows of length \code{(max(y) - min(y)) * y_fraction}.
+#' Samples within the same bin belong to the same "neighbourhood".
+#' @param scale logical. When set to \code{TRUE} x-coordinate widths across all
+#' groups are scaled based on the densiest are in the plot. Default: \code{TRUE}
+#' @param neighbour_limit if the samples within the same y-axis bin are more
+#' than neighbour_limit, the samples's X coordinates will be adjusted.
+#' @param method choose the method to spread the samples within the same
+#' neighbourhood along the x-axis. Available methods: "density",
+#' "neighbourhood" (can be abbreciated, e.g. "d"). See \code{Details}.
+#' @param adjust adjusts the bandwidth of the density kernel when
+#' \code{method == "density"} (see \code{\link[stats]{density}}) or the spread
+#' of the samples within the same group along the x-axis.
+#' @export
+stat_sina <-function(mapping = NULL, data = NULL,
+                     geom = "point", position = "identity",
+                     ...,
+                     binwidth = 0.02,
+                     scale = TRUE,
+                     neighbour_limit = 1,
+                     method = "density",
+                     adjust = 1,
+                     na.rm = FALSE,
+                     show.legend = NA,
+                     inherit.aes = TRUE) {
+  method <- match.arg(method, c("density", "neighbourhood"))
+
+  layer(
+    data = data,
+    mapping = mapping,
+    stat = StatSina,
+    geom = geom,
+    position = position,
+    show.legend = show.legend,
+    inherit.aes = inherit.aes,
+    params = list(
+      binwidth = binwidth,
+      scale = scale,
+      neighbour_limit = neighbour_limit,
+      method = method,
+      adjust = adjust,
+      na.rm = na.rm,
+      ...
+      )
+  )
+}
+
+
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
+#' @export
+StatSina <- ggproto("StatSina", Stat,
+
+  required_aes = c("x", "y"),
+
```

Comment:
Add this for the computed variable
```R
default_aes = aes(x= ..scaled..),
```

For consistency (see `scale_density`) and see comment below why I used 'scaled'